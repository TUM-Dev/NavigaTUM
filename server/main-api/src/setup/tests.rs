use meilisearch_sdk::client::Client;
use sqlx::postgres::PgPoolOptions;
use sqlx::{Pool, Postgres};
use testcontainers::core::{ContainerPort, WaitFor};
use testcontainers::{ContainerAsync, GenericImage, ImageExt};
use testcontainers_modules::{postgres, testcontainers::runners::AsyncRunner};

#[cfg(not(feature = "skip_db_setup"))]
pub struct PostgresTestContainer {
    _container: ContainerAsync<postgres::Postgres>,
    pub pool: Pool<Postgres>,
}

#[cfg(not(feature = "skip_db_setup"))]
impl PostgresTestContainer {
    /// Create a postgres instance for testing against
    pub async fn new() -> Self {
        let container = postgres::Postgres::default()
            .with_tag("16")
            .start()
            .await
            .unwrap();
        let connection_string = format!(
            "postgres://postgres:postgres@{host}:{port}/postgres",
            host = container.get_host().await.unwrap(),
            port = container.get_host_port_ipv4(5432).await.unwrap(),
        );
        let pool = PgPoolOptions::new()
            .connect(&connection_string)
            .await
            .unwrap();
        crate::setup::database::setup(&pool).await.unwrap();
        Self {
            _container: container,
            pool,
        }
    }
}

#[cfg(not(feature = "skip_ms_setup"))]
pub struct MeiliSearchTestContainer {
    _container: ContainerAsync<GenericImage>,
    pub client: Client,
}

#[cfg(not(feature = "skip_ms_setup"))]
impl MeiliSearchTestContainer {
    /// Create a meilisearch instance for testing against
    pub async fn new() -> Self {
        let container = GenericImage::new("getmeili/meilisearch", "v1.8.0")
            .with_exposed_port(ContainerPort::Tcp(7700))
            .with_wait_for(WaitFor::message_on_stderr(
                "Actix runtime found; starting in Actix runtime",
            ))
            .start()
            .await
            .unwrap();
        let meili_url = format!(
            "http://{host}:{port}",
            host = container.get_host().await.unwrap(),
            port = container.get_host_port_ipv4(7700).await.unwrap(),
        );

        let client = Client::new(meili_url.clone(), None::<String>).unwrap();
        super::meilisearch::setup(&client).await.unwrap();
        Self {
            _container: container,
            client,
        }
    }
}

#[tokio::test]
#[cfg(not(feature = "skip_db_setup"))]
async fn test_db_setup() {
    crate::setup_logging();
    let pg = PostgresTestContainer::new().await;
    crate::setup::database::load_data(&pg.pool).await.unwrap();
}

#[tokio::test]
#[cfg(not(feature = "skip_ms_setup"))]
async fn test_meilisearch_setup() {
    crate::setup_logging();
    let ms = MeiliSearchTestContainer::new().await;
    crate::setup::meilisearch::load_data(&ms.client)
        .await
        .unwrap();
}
