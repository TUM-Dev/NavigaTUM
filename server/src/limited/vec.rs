use std::fmt;
use std::vec::IntoIter;

use serde::{Deserialize, Serialize};

use crate::limited::OrMore;

#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct LimitedVec<T>(pub Vec<T>);

impl<T> AsRef<[T]> for LimitedVec<T> {
    fn as_ref(&self) -> &[T] {
        &self.0
    }
}

impl<T> IntoIterator for LimitedVec<T> {
    type Item = T;
    type IntoIter = IntoIter<T>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}

impl<T> LimitedVec<T> {
    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
    pub fn len(&self) -> usize {
        self.0.len()
    }
    pub fn pop(&mut self) -> Option<T> {
        self.0.pop()
    }
}

impl<T> From<Vec<T>> for LimitedVec<T> {
    fn from(value: Vec<T>) -> Self {
        LimitedVec(value)
    }
}

const LIMIT: usize = 3;
impl<T: fmt::Debug> fmt::Debug for LimitedVec<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.0.len() <= LIMIT {
            f.debug_list().entries(self.0.iter().take(LIMIT)).finish()
        } else {
            f.debug_list()
                .entries(
                    self.0
                        .iter()
                        .take(LIMIT)
                        .map(OrMore::Value)
                        .chain([OrMore::More]),
                )
                .finish()
        }
    }
}
impl<T> FromIterator<T> for LimitedVec<T> {
    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
        let mut c = Vec::new();

        for i in iter {
            c.push(i);
        }

        LimitedVec(c)
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_limited_output() {
        let w: LimitedVec<u32> = LimitedVec(vec![]);
        assert_eq!(format!("{w:?}"), "[]");
        let w = LimitedVec(vec![1]);
        assert_eq!(format!("{w:?}"), "[1]");
        let w = LimitedVec(vec![1, 2]);
        assert_eq!(format!("{w:?}"), "[1, 2]");
        let w = LimitedVec(vec![1, 2, 3]);
        assert_eq!(format!("{w:?}"), "[1, 2, 3]");
        let w = LimitedVec(vec![1, 2, 3, 4]);
        assert_eq!(format!("{w:?}"), "[1, 2, 3, ...]");
        let w = LimitedVec(vec![1, 2, 3, 4, 5]);
        assert_eq!(format!("{w:?}"), "[1, 2, 3, ...]");
    }
}
