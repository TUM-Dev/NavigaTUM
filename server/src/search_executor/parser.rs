use logos::Logos;
use std::collections::HashSet;
use std::fmt::{Debug, Formatter};
use tracing::warn;

use super::lexer::Token;

#[derive(Clone, Default, PartialEq, Eq)]
pub struct Filter {
    parents: HashSet<String>,
    types: HashSet<String>,
    usages: HashSet<String>,
}
impl Filter {
    pub fn as_meilisearch_filters(&self) -> String {
        let mut filters = vec![];
        if !self.parents.is_empty() {
            let parents: Vec<&str> = self.parents.iter().map(String::as_str).collect();
            filters.push(format!(
                "((parent_keywords IN {parents:?}) OR (parent_building_names IN {parents:?}) OR (campus IN {parents:?}))"
            ));
        }
        if !self.types.is_empty() {
            let types: Vec<&str> = self.types.iter().map(String::as_str).collect();
            filters.push(format!("(type IN {types:?})"));
        }
        if !self.usages.is_empty() {
            let usages: Vec<&str> = self.usages.iter().map(String::as_str).collect();
            filters.push(format!("(usage IN {usages:?})"));
        }
        filters.join(" AND ")
    }
    pub fn is_empty(&self) -> bool {
        self.parents.is_empty() && self.types.is_empty() && self.usages.is_empty()
    }
}

impl Debug for Filter {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut base = f.debug_struct("Filter");
        if !self.parents.is_empty() {
            base.field("parents", &self.parents);
        }
        if !self.types.is_empty() {
            base.field("types", &self.parents);
        }
        if !self.usages.is_empty() {
            base.field("usages", &self.parents);
        }
        base.finish()
    }
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Sorting {
    location: HashSet<String>,
}

impl Sorting {
    pub fn as_meilisearch_sorting(&self) -> Vec<String> {
        self.location
            .iter()
            .map(|s| format!("_geoPoint({s}):asc"))
            .collect()
    }
    pub fn is_empty(&self) -> bool {
        self.location.is_empty()
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TextToken {
    Text(String),
    SplittableText((String, String)),
}

#[derive(Clone, Default, PartialEq, Eq)]
pub struct ParsedQuery {
    pub tokens: Vec<TextToken>,
    pub filters: Filter,
    pub sorting: Sorting,
}

impl Debug for ParsedQuery {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut base = f.debug_struct("ParsedQuery");
        base.field("tokens", &self.tokens);
        if !self.filters.is_empty() {
            base.field("from", &self.filters);
        }
        if !self.sorting.is_empty() {
            base.field("from", &self.sorting);
        }
        base.finish()
    }
}

impl ParsedQuery {
    pub fn relevant_enough_for_room_highligting(&self) -> bool {
        if self.tokens.len() == 1 {
            return true;
        }
        match self.tokens.first() {
            Some(first) => match first {
                TextToken::Text(t) => t.len() > 3,
                TextToken::SplittableText((t0, _)) => t0.len() > 3,
            },
            None => false,
        }
    }
}

impl From<&str> for ParsedQuery {
    fn from(query: &str) -> Self {
        let mut result = Self::default();
        for token in Token::lexer(query) {
            match token {
                Ok(Token::Text(s)) => {
                    result.tokens.push(TextToken::Text(s));
                }
                Ok(Token::SplittableText((s1, s2))) => {
                    result.tokens.push(TextToken::SplittableText((s1, s2)));
                }
                Ok(Token::ParentFilter(filter)) => {
                    result.filters.parents.insert(filter);
                }
                Ok(Token::UsageFilter(filter)) => {
                    result.filters.usages.insert(filter);
                }
                Ok(Token::TypeFilter(filter)) => {
                    result.filters.types.insert(filter);
                }
                Ok(Token::LocationSort(location)) => {
                    result.sorting.location.insert(location);
                }
                Err(_) => {
                    warn!("Error in query parsing");
                }
            }
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::*;

    #[test]
    fn parent_filter() {
        for filter in ["in:", "@"] {
            for sep in ["", " "] {
                assert_eq!(
                    ParsedQuery::from(format!("{filter}{sep}foo").as_str()).filters,
                    Filter {
                        parents: HashSet::from(["foo".to_string()]),
                        ..Default::default()
                    }
                );
            }
        }
    }

    #[test]
    fn usage_filters() {
        for filter in ["usage:", "nutzung:", "="] {
            for sep in ["", " "] {
                assert_eq!(
                    ParsedQuery::from(format!("{filter}{sep}foo").as_str()).filters,
                    Filter {
                        usages: HashSet::from(["foo".to_string()]),
                        ..Default::default()
                    }
                );
            }
        }
    }

    #[test]
    fn type_filters() {
        for sep in ["", " "] {
            assert_eq!(
                ParsedQuery::from(format!("type:{sep}foo").as_str()).filters,
                Filter {
                    types: HashSet::from(["foo".to_string()]),
                    ..Default::default()
                }
            );
        }
    }

    #[test]
    fn location_sort() {
        for sep in ["", " "] {
            assert_eq!(
                ParsedQuery::from(format!("near:{sep}45.32,59.3").as_str()).sorting,
                Sorting {
                    location: HashSet::from(["45.32,59.3".to_string()]),
                }
            );
        }
        for sep in ["", " "] {
            assert_eq!(
                ParsedQuery::from(format!("near:{sep}45.3,59.00000003").as_str()).sorting,
                Sorting {
                    location: HashSet::from(["45.3,59.00000003".to_string()]),
                }
            );
        }
    }

    #[test]
    fn text_token() {
        assert_eq!(
            ParsedQuery::from("foo").tokens,
            vec![TextToken::Text("foo".to_string())]
        );
        assert_eq!(
            ParsedQuery::from("foo foo").tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::Text("foo".to_string()),
            ]
        );
        assert_eq!(
            ParsedQuery::from("foo hs1").tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::SplittableText(("hs".to_string(), "1".to_string())),
            ]
        );
    }

    #[test]
    fn text_filter_mixed() {
        assert_eq!(
            ParsedQuery::from("foo in:abc bar @abc foo near:45.32,59.3").tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::Text("bar".to_string()),
                TextToken::Text("foo".to_string()),
            ]
        );
        assert_eq!(
            ParsedQuery::from(
                "foo in:abc bar @abc =def usage:dd nutzung:gh type:fdh foo near:45.32,59.3"
            )
            .tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::Text("bar".to_string()),
                TextToken::Text("foo".to_string()),
            ]
        );
    }
}
