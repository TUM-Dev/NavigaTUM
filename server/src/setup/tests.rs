use meilisearch_sdk::client::Client;
use testcontainers_modules::testcontainers::{ContainerAsync, ImageExt};
use testcontainers_modules::{meilisearch, testcontainers::runners::AsyncRunner};

#[cfg(feature = "test-with-geodata")]
pub struct PostgresTestContainer {
    _container: ContainerAsync<testcontainers_modules::postgres::Postgres>,
    pub pool: sqlx::Pool<sqlx::Postgres>,
}
#[cfg(feature = "test-with-geodata")]

impl PostgresTestContainer {
    /// Create a postgres instance for testing against
    pub async fn new() -> Self {
        let container = postgres::Postgres::default()
            .with_tag("16-3.4")
            .with_name("postgis/postgis")
            .start()
            .await
            .unwrap();
        let connection_string = format!(
            "postgres://postgres:postgres@{host}:{port}/postgres",
            host = container.get_host().await.unwrap(),
            port = container.get_host_port_ipv4(5432).await.unwrap(),
        );
        let pool = sqlx::postgres::PgPoolOptions;
        ::new().connect(&connection_string).await.unwrap();
        crate::setup::database::setup(&pool).await.unwrap();
        Self {
            _container: container,
            pool,
        }
    }
}

pub struct MeiliSearchTestContainer {
    _container: ContainerAsync<meilisearch::Meilisearch>,
    pub client: Client,
}

impl MeiliSearchTestContainer {
    /// Create a meilisearch instance for testing against
    pub async fn new() -> Self {
        let container = meilisearch::Meilisearch::default()
            .with_tag("v1.9.0")
            .start()
            .await
            .unwrap();
        let meili_url = format!(
            "http://{host}:{port}",
            host = container.get_host().await.unwrap(),
            port = container.get_host_port_ipv4(7700).await.unwrap(),
        );

        let client = Client::new(meili_url.clone(), None::<String>).unwrap();
        super::meilisearch::setup(&client).await.unwrap();
        Self {
            _container: container,
            client,
        }
    }
}

#[cfg(feature = "test-with-geodata")]
#[tokio::test]
#[ignore]
#[tracing_test::traced_test]
async fn test_db_setup() {
    let pg = PostgresTestContainer::new().await;
    for i in 0..20 {
        let res = crate::setup::database::load_data(&pg.pool).await;
        if let Err(e) = res {
            tracing::error!("failed to load db because {e:?}. Retrying for 20s");
            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
        } else {
            tracing::info!("successfully initalised the db in try {i}");
            break;
        }
    }
}

#[cfg(feature = "test-with-geodata")]
#[tokio::test]
#[tracing_test::traced_test]
async fn test_meilisearch_setup() {
    let ms = MeiliSearchTestContainer::new().await;
    crate::setup::meilisearch::load_data(&ms.client)
        .await
        .unwrap();
}
