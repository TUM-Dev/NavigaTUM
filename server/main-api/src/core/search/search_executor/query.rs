use crate::core::search::search_executor::preprocess;
use crate::core::search::SanitisedSearchQueryArgs;
use meilisearch_sdk::errors::Error;
use meilisearch_sdk::indexes::Index;
use meilisearch_sdk::search::{MultiSearchResponse, SearchQuery, Selectors};
use meilisearch_sdk::Client;
use serde::Deserialize;

fn ms_url() -> String {
    // meilisearch should not be a public service as by their docs
    format!(
        "http://{}:{}/",
        std::env::var("MIELISEARCH_HOST").unwrap_or_else(|_| "localhost".to_string()),
        std::env::var("MIELISEARCH_PORT").unwrap_or_else(|_| "7700".to_string())
    )
}

#[derive(Deserialize, Clone, Debug)]
#[allow(dead_code)]
pub(super) struct MSHit {
    ms_id: String,
    pub(super) id: String,
    pub(super) name: String,
    pub(super) arch_name: Option<String>,
    pub(super) r#type: String,
    pub(super) type_common_name: String,
    pub(super) parent_building_names: Vec<String>,
    parent_keywords: Vec<String>,
    pub(super) campus: Option<String>,
    address: Option<String>,
    usage: Option<String>,
    rank: i32,
}

pub(super) struct GeoEntryQuery {
    parsed_input: preprocess::SearchInput,
    args: SanitisedSearchQueryArgs,
    highlighting: (String, String),
}

impl GeoEntryQuery {
    pub fn from(
        parsed_input: &preprocess::SearchInput,
        args: &SanitisedSearchQueryArgs,
        highlighting: &(String, String),
    ) -> Self {
        Self {
            parsed_input: parsed_input.clone(),
            args: *args,
            highlighting: highlighting.clone(),
        }
    }
    pub async fn execute(self) -> Result<MultiSearchResponse<MSHit>, Error> {
        let q_default = self.parsed_input.to_default_query();
        let client = Client::new(ms_url(), std::env::var("MEILI_MASTER_KEY").ok());
        let entries = client.index("entries");

        // Currently ranking is designed to put buildings at the top if they equally
        // match the term compared to a room. For this reason there is only a search
        // for all entries and only rooms, search matching (and relevant) buildings can be
        // expected to be at the top of the merged search. However sometimes a lot of
        // buildings will be hidden (e.g. building parts), so the extra room search ....
        client
            .multi_search()
            .with_search_query(self.merged_query(&entries, &q_default))
            .with_search_query(self.buildings_query(&entries, &q_default))
            .with_search_query(self.rooms_query(&entries, &self.parsed_input.to_room_query()))
            .execute::<MSHit>()
            .await
    }

    fn merged_query<'a>(&'a self, entries: &'a Index, query: &'a str) -> SearchQuery<'a> {
        SearchQuery::new(entries)
            .with_facets(Selectors::Some(&["facet"]))
            .with_highlight_pre_tag(&self.highlighting.0)
            .with_highlight_post_tag(&self.highlighting.1)
            .with_attributes_to_highlight(Selectors::Some(&["name"]))
            .with_query(query)
            .with_limit(self.args.limit_all)
            .build()
    }

    fn buildings_query<'a>(&'a self, entries: &'a Index, query: &'a str) -> SearchQuery<'a> {
        SearchQuery::new(entries)
            .with_facets(Selectors::Some(&["facet"]))
            .with_highlight_pre_tag(&self.highlighting.0)
            .with_highlight_post_tag(&self.highlighting.1)
            .with_attributes_to_highlight(Selectors::Some(&["name"]))
            .with_query(query)
            .with_limit(2 * self.args.limit_buildings) // we might do reordering later
            .with_filter("facet = \"building\"")
            .build()
    }

    fn rooms_query<'a>(&'a self, entries: &'a Index, query: &'a str) -> SearchQuery<'a> {
        SearchQuery::new(entries)
            .with_facets(Selectors::Some(&["facet"]))
            .with_highlight_pre_tag(&self.highlighting.0)
            .with_highlight_post_tag(&self.highlighting.1)
            .with_attributes_to_highlight(Selectors::Some(&["name"]))
            .with_query(query)
            .with_limit(self.args.limit_rooms)
            .with_filter("facet = \"room\"")
            .build()
    }
}
