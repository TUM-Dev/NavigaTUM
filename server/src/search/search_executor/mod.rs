use meilisearch_sdk::client::Client;
use serde::Serialize;
use tracing::error;

use crate::limited::vec::LimitedVec;
use crate::search::search_executor::parser::ParsedQuery;
use crate::search::search_executor::query::MSHit;

use super::{Highlighting, Limits};

mod formatter;
mod lexer;
mod merger;
mod parser;
mod query;

#[derive(Serialize, Debug, Clone)]
pub struct ResultsSection {
    facet: String,
    entries: Vec<ResultEntry>,
    n_visible: usize,
    #[serde(rename = "estimatedTotalHits")]
    estimated_total_hits: usize,
}

#[derive(Serialize, Default, Debug, Clone)]
struct ResultEntry {
    #[serde(skip)]
    hit: MSHit,
    id: String,
    r#type: String,
    name: String,
    subtext: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    subtext_bold: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    parsed_id: Option<String>,
}
#[tracing::instrument(skip(client))]
pub async fn do_geoentry_search(
    client: &Client,
    q: String,
    highlighting: Highlighting,
    limits: Limits,
) -> LimitedVec<ResultsSection> {
    let parsed_input = ParsedQuery::from(q.as_str());

    match query::GeoEntryQuery::from((client, &parsed_input, &limits, &highlighting))
        .execute()
        .await
    {
        Ok(response) => {
            let (section_buildings, mut section_rooms) = merger::merge_search_results(
                &limits,
                response.results.first().unwrap(),
                response.results.get(1).unwrap(),
                response.results.get(2).unwrap(),
            );
            let visitor = formatter::RoomVisitor::from((parsed_input, highlighting));
            section_rooms
                .entries
                .iter_mut()
                .for_each(|r| visitor.visit(r));

            match section_buildings.n_visible {
                0 => LimitedVec(vec![section_rooms, section_buildings]),
                _ => LimitedVec(vec![section_buildings, section_rooms]),
            }
        }
        Err(e) => {
            // error should be serde_json::error
            error!("Error searching for results: {e:?}");
            LimitedVec(vec![])
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::setup::tests::MeiliSearchTestContainer;
    use std::fmt::{Display, Formatter};

    #[derive(serde::Deserialize)]
    struct TestQuery {
        target: String,
        query: String,
        among: Option<usize>,
        comment: Option<String>,
    }

    impl TestQuery {
        fn load_good() -> Vec<Self> {
            serde_yaml::from_str(include_str!("test-queries.good.yaml")).unwrap()
        }
        fn load_bad() -> Vec<Self> {
            serde_yaml::from_str(include_str!("test-queries.bad.yaml")).unwrap()
        }
        fn actual_matches_among(&self, actual: &[ResultsSection]) -> bool {
            let among = self.among.unwrap_or(1);
            let mut acceptable_range = actual.iter().flat_map(|r| r.entries.clone()).take(among);
            acceptable_range.any(|r| r.id == self.target)
        }
        async fn search(&self, client: &Client) -> Vec<ResultsSection> {
            do_geoentry_search(
                client,
                self.query.clone(),
                Highlighting::default(),
                Limits::default(),
            )
            .await
            .0
        }
    }
    impl Display for TestQuery {
        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
            write!(
                f,
                "'{query}' should get '{target}' in {among} results",
                query = self.query,
                target = self.target,
                among = self.among.unwrap_or(1),
            )?;
            if let Some(comment) = &self.comment {
                write!(f, " # {comment}")?;
            }
            Ok(())
        }
    }
    #[tokio::test]
    #[tracing_test::traced_test]
    async fn test_good_queries() {
        let ms = MeiliSearchTestContainer::new().await;
        crate::setup::meilisearch::load_data(&ms.client)
            .await
            .unwrap();
        for query in TestQuery::load_good() {
            let actual = query.search(&ms.client).await;
            assert!(
                query.actual_matches_among(&actual),
                "{query}\n\
                Since it can't, please move it to .bad list"
            );

            insta::with_settings!({
                info => &format!("{query}"),
                description => query.comment.unwrap_or_default(),
            }, {
                        insta::assert_yaml_snapshot!(actual);
            });
        }
    }

    #[tokio::test]
    #[tracing_test::traced_test]
    async fn test_bad_queries() {
        let ms = MeiliSearchTestContainer::new().await;
        crate::setup::meilisearch::load_data(&ms.client)
            .await
            .unwrap();
        for query in TestQuery::load_bad() {
            let actual = query.search(&ms.client).await;
            assert!(
                !query.actual_matches_among(&actual),
                "{query}\n\
                Since it can, please move it to .good list"
            );

            insta::with_settings!({
                info => &format!("{query}"),
                description => query.comment.unwrap_or_default(),
            }, {
                insta::assert_yaml_snapshot!(actual);
            });
        }
    }
}
