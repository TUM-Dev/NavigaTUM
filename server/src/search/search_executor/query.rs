use meilisearch_sdk::client::Client;
use meilisearch_sdk::errors::Error;
use meilisearch_sdk::indexes::Index;
use meilisearch_sdk::search::{MultiSearchResponse, SearchQuery, Selectors};
use serde::Deserialize;
use std::fmt::{Debug, Formatter};

use crate::search::search_executor::parser::{Filter, ParsedQuery, TextToken};
use crate::search::{Highlighting, Limits};

#[derive(Deserialize, Default, Clone)]
#[allow(dead_code)]
pub(super) struct MSHit {
    ms_id: String,
    pub(super) id: String,
    pub(super) name: String,
    pub(super) arch_name: Option<String>,
    pub(super) r#type: String,
    pub(super) type_common_name: String,
    pub(super) parent_building_names: Vec<String>,
    parent_keywords: Vec<String>,
    pub(super) campus: Option<String>,
    address: Option<String>,
    usage: Option<String>,
    rank: i32,
}
impl Debug for MSHit {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("MSHit")
            .field("id", &self.id)
            .field("name", &self.name)
            .finish()
    }
}

struct GeoEntryFilters {
    default: String,
    rooms: String,
    buildings: String,
}
impl Debug for GeoEntryFilters {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut base = f.debug_struct("GeoEntryFilters");
        if !self.default.is_empty() {
            base.field("default", &self.default);
        }
        if !self.rooms.is_empty() {
            base.field("rooms", &self.rooms);
        }
        if !self.buildings.is_empty() {
            base.field("buildings", &self.buildings);
        }
        base.finish()
    }
}

impl From<&Filter> for GeoEntryFilters {
    fn from(filters: &Filter) -> Self {
        let ms_filter = filters.as_meilisearch_filters();
        let separator = if ms_filter.is_empty() { " " } else { " AND " };
        Self {
            default: ms_filter.clone(),
            buildings: format!("facet = \"building\"{separator}{ms_filter}"),
            rooms: format!("facet = \"room\"{separator}{ms_filter}"),
        }
    }
}

pub(super) struct GeoEntryQuery {
    client: Client,
    parsed_input: ParsedQuery,
    limits: Limits,
    highlighting: Highlighting,
    filters: GeoEntryFilters,
    sorting: Vec<String>,
}

impl Debug for GeoEntryQuery {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut base = f.debug_struct("GeoEntryQuery");
        base.field("parsed_input", &self.parsed_input)
            .field("limits", &self.limits)
            .field("highlighting", &self.highlighting)
            .field("filters", &self.filters);
        if !self.sorting.is_empty() {
            base.field("sorting", &self.sorting);
        }
        base.finish()
    }
}

impl From<(&Client, &ParsedQuery, &Limits, &Highlighting)> for GeoEntryQuery {
    fn from(
        (client, parsed_input, limits, highlighting): (
            &Client,
            &ParsedQuery,
            &Limits,
            &Highlighting,
        ),
    ) -> Self {
        Self {
            client: client.clone(),
            parsed_input: parsed_input.clone(),
            limits: *limits,
            highlighting: highlighting.clone(),
            filters: GeoEntryFilters::from(&parsed_input.filters),
            sorting: parsed_input.sorting.as_meilisearch_sorting(),
        }
    }
}

impl GeoEntryQuery {
    pub async fn execute(self) -> Result<MultiSearchResponse<MSHit>, Error> {
        let q_default = self.prompt_for_querying();
        let entries = self.client.index("entries");

        // due to lifetime shenanigans this is added here (I can't make it move down to the other statements)
        // If you can make it, please propose a PR, I know that this is really hacky ^^
        let sorting = self
            .sorting
            .iter()
            .map(String::as_str)
            .collect::<Vec<&str>>();

        // Currently ranking is designed to put buildings at the top if they equally
        // match the term compared to a room. For this reason there is only a search
        // for all entries and only rooms, search matching (and relevant) buildings can be
        // expected to be at the top of the merged search. However sometimes a lot of
        // buildings will be hidden (e.g. building parts), so the extra room search ....
        self.client
            .multi_search()
            .with_search_query(
                self.merged_query(&entries, &q_default)
                    .with_sort(&sorting)
                    .build(),
            )
            .with_search_query(
                self.buildings_query(&entries, &q_default)
                    .with_sort(&sorting)
                    .build(),
            )
            .with_search_query(
                self.rooms_query(&entries, &self.prompt_for_querying_room())
                    .with_sort(&sorting)
                    .build(),
            )
            .execute::<MSHit>()
            .await
    }

    fn prompt_for_querying(&self) -> String {
        self.parsed_input
            .tokens
            .clone()
            .into_iter()
            .map(|s| match s {
                TextToken::Text(t) => t,
                TextToken::SplittableText((t1, t2)) => format!("{t1}{t2}"),
            })
            .collect::<Vec<String>>()
            .join(" ")
    }
    fn prompt_for_querying_room(&self) -> String {
        self.parsed_input
            .tokens
            .clone()
            .into_iter()
            .map(|s| match s {
                TextToken::Text(t) => t,
                TextToken::SplittableText((t1, t2)) => format!("{t1} {t2} {t1}{t2}"),
            })
            .collect::<Vec<String>>()
            .join(" ")
    }

    fn common_query<'b: 'a, 'a>(
        &'b self,
        entries: &'a Index,
    ) -> SearchQuery<'a, meilisearch_sdk::DefaultHttpClient> {
        SearchQuery::new(entries)
            .with_facets(Selectors::Some(&["facet"]))
            .with_highlight_pre_tag(&self.highlighting.pre)
            .with_highlight_post_tag(&self.highlighting.post)
            .with_attributes_to_highlight(Selectors::Some(&["name"]))
            .build()
    }

    fn merged_query<'a>(
        &'a self,
        entries: &'a Index,
        query: &'a str,
    ) -> SearchQuery<'a, meilisearch_sdk::DefaultHttpClient> {
        let mut s = self
            .common_query(entries)
            .with_query(query)
            .with_limit(self.limits.total_count)
            .build();
        if !self.filters.default.is_empty() {
            s = s.with_filter(&self.filters.default).build();
        }
        s
    }

    fn buildings_query<'a>(
        &'a self,
        entries: &'a Index,
        query: &'a str,
    ) -> SearchQuery<'a, meilisearch_sdk::DefaultHttpClient> {
        self.common_query(entries)
            .with_query(query)
            .with_limit(2 * self.limits.buildings_count) // we might do reordering later
            .with_filter(&self.filters.buildings)
            .build()
    }

    fn rooms_query<'a>(
        &'a self,
        entries: &'a Index,
        query: &'a str,
    ) -> SearchQuery<'a, meilisearch_sdk::DefaultHttpClient> {
        self.common_query(entries)
            .with_query(query)
            .with_limit(self.limits.rooms_count)
            .with_filter(&self.filters.rooms)
            .build()
    }
}
