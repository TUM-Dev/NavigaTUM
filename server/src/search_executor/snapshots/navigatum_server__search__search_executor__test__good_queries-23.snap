---
source: src/search/search_executor/mod.rs
description: ""
expression: actual
info: "'342 Physik' should get '5101.EG.342' in 1 results"
---
- facet: rooms
  entries:
    - id: 5101.EG.342
      type: room
      name: "5101.EG.\u0019342\u0017 (\u0019Physik\u0017labor)"
      subtext: "garching, Physik I"
      subtext_bold: 2342@5101
    - id: 5101.01.342
      type: room
      name: "5101.01.\u0019342\u0017 (Sitzungs-,  Konferenzraum)"
      subtext: "garching, Physik I"
      subtext_bold: 3342@5101
    - id: 4213.01.342
      type: room
      name: "4213.01.\u0019342\u0017 (Labor (gem. Nutzung 1124302010;1110064600))"
      subtext: "weihenstephan, Lebensmitteltechnikum"
      subtext_bold: O42@4213
    - id: 4213.EG.342
      type: room
      name: "4213.EG.\u0019342\u0017 (Labor)"
      subtext: "weihenstephan, Lebensmitteltechnikum"
      subtext_bold: E42@4213
    - id: 4317.03.342
      type: room
      name: "4317.03.\u0019342\u0017 (Labor)"
      subtext: "weihenstephan, Tierwissenschaften"
      subtext_bold: 3.42@4317
    - id: 2910.03.342
      type: room
      name: "2910.03.\u0019342\u0017 (Teeküche)"
      subtext: "stammgelände, RiWa 1 (HfP/GOV)"
      subtext_bold: H.342@2910
    - id: 0503.01.342
      type: room
      name: "0503.01.\u0019342\u0017 (Büro)"
      subtext: "stammgelände, Thierschbau (Z3)"
      subtext_bold: 1342@0503
    - id: 0503.03.342
      type: room
      name: "0503.03.\u0019342\u0017 (Büro)"
      subtext: "stammgelände, Thierschbau (Z3)"
      subtext_bold: 3342@0503
    - id: 0503.EG.342
      type: room
      name: "0503.EG.\u0019342\u0017 (Sekretariat 2)"
      subtext: "stammgelände, Thierschbau (Z3)"
      subtext_bold: 0342@0503
    - id: 2334.01.342
      type: room
      name: "2334.01.\u0019342\u0017 ((34.1.406) WC-Herren)"
      subtext: "campus-im-olympiapark-sz, CiO/SG Institute West, Bibliothek"
      subtext_bold: 01.2334.342@2334
  n_visible: 10
  estimatedTotalHits: "[estimatedTotalHits]"
- facet: sites_buildings
  entries: []
  n_visible: 0
  estimatedTotalHits: "[estimatedTotalHits]"
